spinn_utilities/make_tools/convertor.py

```diff
@@ -0,0 +1,147 @@
+
+    def _get_id(self):
+        RANGE_PER_DIR = 1000
+
+        rangefile = os.path.join(RANGE_DIR, "log.ranges")
+        range_start = 0
+        filename = self._dest
+        common_dir = self._find_common_based_on_environ()
+        if filename.startswith(common_dir):
+            filename = filename[len(common_dir)+1:]
+
+        # If the range_file does not exist create it and use range_start
+        if not os.path.exists(rangefile):
+            with open(rangefile, 'w') as log_ranges_file:
+                log_ranges_file.write("AUTOGENERATED DO NOT EDIT\n")
+                log_ranges_file.write("{} {}\n".format(
+                    range_start, filename))
+            return range_start
+
+        # Check if the file is ranged or find highest range so far
+        highest_found = range_start
+        with open(rangefile, 'r') as log_ranges_file:
+            data_lines = iter(log_ranges_file)
+            next(data_lines)  # Ignore do not edit
+            for line in data_lines:
+                parts = line.split(" ", 1)
+                if filename.strip() == parts[1].strip():
+                    return int(parts[0])
+                else:
+                    highest_found = max(highest_found, int(parts[0]))
+
+        # Go one step above best found
+        new_start = highest_found + RANGE_PER_DIR
+
+        # Append to range file in case rebuilt without clean
+        with open(rangefile, 'a') as log_ranges_file:
+            log_ranges_file.write("{} {}\n".format(new_start, filename))
+        return new_start
+
+    def copy_if_newer(self, src_path):
+        destination = self._newer_destination(src_path)
+        if destination is None:
+            return  # newer so no need to copy
+        shutil.copy2(src_path, destination)
+
+    def _any_destination(self, path):
+        # Here we need the local seperator
+        destination = path.replace(
+            os.path.sep + self._src_basename + os.path.sep,
+            os.path.sep + self._dest_basename + os.path.sep)
+        return destination
+
+    def _newer_destination(self, path):
+        destination = self._any_destination(path)
+        if not os.path.exists(destination):
+            return destination
+        # need to floor the time as some copies ignore the partial second
+        src_time = math.floor(os.path.getmtime(path))
+        dest_time = math.floor(os.path.getmtime(destination))
+        if src_time > dest_time:
+            return destination
+        else:
+            # print ("ignoring {}".format(destination))
+            return None
+
+    def _mkdir(self, path):
+        destination = self._any_destination(path)
+        if not os.path.exists(destination):
+            os.mkdir(destination)
+        if not os.path.exists(destination):
+            raise Exception("mkdir failed {}".format(destination))
+
+    def _find_common_based_on_environ(self):
+        if 'SPINN_DIRS' not in os.environ:
```

Comment:
make into magic_param as then if we ever change it, you only need change it in one place. this method has 2 of each already

Reply:
"{} {}\n" Here I disagree with making this a CONSTANT or even a magic. Keep the line needing parameters with the code that provides them.